Angular Forms Directive

DickSmith opened the pull request:
This change allows for the use of `ngModel` and `formControlName` when using the `nativescript-drop-down` with Angular.

- No breaking changes to previous functionality or usage for either Angular or Vanilla NativeScript.
- Compatible with Angular AoT and Webpack.
- Build scripts updated (will only add an additional 3 files to npm package: `angular/index.js`, `angular/index.d.ts`, and `angular/index.metadata.json` for AoT compilation).

To use the new directives remove `registerElement("DropDown", () => require("nativescript-drop-down/drop-down").DropDown);` from the application and import the module as usual in `ngModule`:

```
import { DropDownModule } from “nativescript-drop-down/angular”;
…
@NgModule({
	…
	imports: [
		…
		DropDownModule,
		…
	],
	…
})
```

_Based on `nativescript-angular/value-accessors/selectedIndex-value-accessor.ts` and `nativescript-angular/nativescript-angular/forms.ts` with some changes to conform to project linting._

PeterStaev commented:
@DickSmith , amazing work! Can you please see to also add the module usage to the readme and the demo-ng project, so it is clear to users. 
Also seems you are missing a dev dep to `@angular/forms`.

DickSmith commented:
Per discussion:
- Updated `README` with migration and example usage.
- Fixed error (`Can't resolve all parameters for SelectedIndexValueAccessor (?)`) for non-webpack by adding `@Inject` (AoT/Webpack worked without it, and most testing had been done against Webpack since it is usually more likely to fail.
- Minimally updated `demo-ng` to use `ngModel`
- Updated package to include `@angular/forms` in `devDependencies` (for some reason just adding `core` had also installed `forms` locally which is why I had missed in when testing locally).

Let me know if there’s anything else. I would maybe also add a peerDependency for `tns-core-modules` to be >= to 3.0 to make it clearer to users during npm install that the new plugin will only work once they update, but I’ll leave that up to you. :D

DickSmith commented:
*Fix for unselected state*
Currently the value is being forced to int (based on what I copied from the main `nativescript-angular` repo), which results in null/undefined/empty string being forced to 0, meaning the hint is never shown and there isn’t the possibility to default to an unselected state. (This also is occurring with the ListPicker from the main repo, with the default state always being forced to index 0).

It’s a preference, I think, but since I’m using this with reactive `FormControls` and I often default new fields to an empty string, I added empty string to also result in an unselected/null state, but I can remove it if you have a preference for otherwise (and just set the value to null in own my forms by default), but since this value is an index, it seems to me an empty string should be treated as null rather than 0.

PeterStaev commented on angular/index.ts:
```diff
@@ -43,6 +43,10 @@ export class SelectedIndexValueAccessor extends BaseValueAccessor<SelectableView
 
     public writeValue(value: any): void {
+        if (value === undefined || value === null || value === "") {
+            return null;
```
Shouldn't this and the line right after the `if` be changed to:
```ts
if (value === undefined || value === null ||  value === "") {
   this._normalizedValue = null;
}
else {
  this._normalizedValue = convertToInt(value);
}
```
And then leave the rest of logic to execute even if the value is null or empty. This way if the form value is reset to null or empty the `selectedIndex` property of the drop down will be set to null and thus show the hint?

PeterStaev commented on uglify-mangle-excludes.ts:
```diff
@@ -10,5 +10,9 @@ export default [
-    "DropDownTouchListener"
+    "DropDownTouchListener",
+
+    // Angular
```
Also do we need these angular specific items to be excluded from mangling? From what I see none of the similar things in the `nativescript-angular` are excluded.

DickSmith replied:
Very possibly not, I actually don't know much about uglify, but saw that you had added all the other named classes in the repo, so assumed it was the same case as Android and Proguard exclusions.

DickSmith commented:
Removing Angular from uglify excludes.

Moving initial `setSelection` for Android to `initNativeView` to solve error in some contexts of `Cannot read property ‘setSelection’ of undefined, undefined`.

PeterStaev commented:
Once again thanks for your awesome work @DickSmith ! Published the new RC version on npm. 

Make sure you contact Jen Looper on the Nativescript Slack channel to get a {N} swag for your contribution 👏
